testbenches/dpi/bsg_nonsynth_dpi_manycore.hpp

```diff
@@ -0,0 +1,175 @@
+                        : dpi_base(hierarchy),
+                          d2f_req(hierarchy + ".d2f_req_i"),
+                          f2d_rsp(hierarchy + ".f2d_rsp_i"),
+                          f2d_req(hierarchy + ".f2d_req_i"),
+                          config(hierarchy + ".rom")
+                {
+                        prev = svSetScope(scope);
+                        max_credits = bsg_dpi_credits_get_max();
+                        svSetScope(prev);
+                }
+
+                /**
+                 * Get the number of manycore credits currently
+                 * available to the bsg_nonsynth_dpi_manycore
+                 * instance.
+                 *
+                 * @param[out] credits  The number of manycore network credits available
+                 *
+                 * @return BSG_NONSYNTH_DPI_SUCCESS on success,
+                 * BSG_NONSYNTH_DPI_NOT_WINDOW when not in valid clock
+                 * window.
+                 */
+                int get_credits(int& credits){
+                        prev = svSetScope(scope);
+                        if(!bsg_dpi_credits_is_window()){
+                                svSetScope(prev);
+                                return BSG_NONSYNTH_DPI_NOT_WINDOW;
+                        }
+
+                        credits = bsg_dpi_credits_get_cur();
+                        svSetScope(prev);
+                        return BSG_NONSYNTH_DPI_SUCCESS;
+                }
+
+                /**
+                 * Transmit a request packet onto the manycore network
+                 * using the DPI interface.
+                 *
+                 * @param[in] data   Padded packet data to
+                 *   transmit. (The module will handle formatting)
+                 *
+                 * @return BSG_NONSYNTH_DPI_SUCCESS on success
+                 *         (Recoverable Errors)
+                 *         BSG_NONSYNTH_DPI_NOT_WINDOW when not in valid clock window
+                 *         BSG_NONSYNTH_DPI_NO_CREDITS when no transmit credits are available
+                 *         BSG_NONSYNTH_DPI_NOT_READY when the packet was not transmitted (call again next cycle)
+                 *         (Unrecoverable Errors)
+                 *         BSG_NONSYNTH_DPI_INVALID when the number of credits available makes no sense
+                 */
+                int tx_req(const __m128i &data){
+                        int res = BSG_NONSYNTH_DPI_SUCCESS;
+
+                        // Get credits checks for valid window
+                        if(!cur_credits)
+                                res = get_credits(cur_credits);
+
+                        if(res != BSG_NONSYNTH_DPI_SUCCESS)
+                                return res;
+
+                        if(cur_credits == 0)
+                                return BSG_NONSYNTH_DPI_NO_CREDITS;
+
+                        if(cur_credits < 0)
+                                return BSG_NONSYNTH_DPI_INVALID;
```

Comment:
is `BSG_NONSYNTH_DPI_INVALID` supposed so signify that the input is bad?

Reply:
No, it signals that cur_credits is an invalid value

Reply:
In that case this is the wrong error code. Maybe you just want to return FAIL.

Reply:
https://github.com/bespoke-silicon-group/basejump_stl/blob/eb8140c080f068f397142aea8e697e7f856037e6/bsg_test/bsg_nonsynth_dpi_errno.hpp#L55

Reply:
Or maybe you want not valid. The string for that error code should be more specific e.g. "Output not valid"

Reply:
Actually - what you *really* should do is print an error message saying "hey - the DPI went off the rails at this spot" so that if it ever does happen we know where in the code to look. Then return FAIL, because realistically there's not much to be done about it at runtime.

Reply:
Looking at the code, negative credits can't happen. If a credit error happens, it triggers a separate set of warnings in bsg_manycore_endpoint_standard.

Reply:
I've removed this check